File: src/main/java/simulations/PlatonSpace.java
package simulations;

import java.awt.*;
import spacesimulation.*;
import spacesimulation.algebra.*;

public class PlatonSpace extends Simulation {
    private Point3d[] points;
    private Vec[] forces;
    private final int RADIUS = 1000;
    CartesianCoordinateSystem coordSys = new CartesianCoordinateSystem(true, RADIUS * 2, RADIUS / 10, new Color(110, 106, 160));
    Color colorLines = new Color(200, 200, 200);
    Color colorPoints = new Color(163, 153, 239);
    Color colorInput = new Color(200, 200, 200);

    public PlatonSpace(int amountOfPoint3ds, Simulator simulator) {
        super(simulator);
        points = new Point3d[amountOfPoint3ds];
        forces = new Vec[amountOfPoint3ds];
        for (int i = 0; i < points.length; i++) points[i] = new Point3d(2 * RADIUS * Math.random() - RADIUS, 2 * RADIUS * Math.random() - RADIUS, 2 * RADIUS * Math.random() - RADIUS);
        for (int i = 0; i < forces.length; i++) forces[i] = new Vec(0, 0, 0);
    }

    public void tick(double dtInSec) {
        calcResultingForceOnPoint3d();
        movePoints();
        keepPointsInOrb();
        drawer.setWindowHeightAndWidth(simulator.getWidth(), simulator.getHeight());
    }

    public void render(Graphics g) {
        coordSys.render(drawer, g);
        drawPoints(g);
        if (simulator.getKeymanager().f) drawForces(g);
    }

    private void drawForces(Graphics g) {
        for (int i = 0; i < points.length; i++) {
            Vec force = forces[i];
            g.setColor(Color.ORANGE);
            double shorten = 1 / force.getLength();
            drawer.drawLine(points[i].x, points[i].y, points[i].z, points[i].x + force.x * shorten, points[i].y + force.y * shorten, points[i].z + force.z * shorten, g);
        }
    }

    private void keepPointsInOrb() {
        for (int i = 0; i < points.length; i++) {
            if (points[i].getPositionVector().getLength() > RADIUS) {
                Vec positionVector = points[i].getPositionVector();
                points[i].set(positionVector.scale(RADIUS / positionVector.getLength()));
            }
        }
    }

    private void movePoints() {
        for (int i = 0; i < points.length; i++) {
            points[i].add(forces[i]);
        }
    }

    private void calcResultingForceOnPoint3d() {
        for (int i = 0; i < points.length; i++) {
            Point3d first = points[i];
            for (Point3d other : points) {
                if (!first.equals(other)) {
                    double scalar = 10000000 / Math.pow(first.getDistanceTo(other), 2);
                    forces[i].add(other.getDirectionTo(first).scale(scalar));
                }
            }
        }
    }

    private void drawPoints(Graphics g) {
        for (Point3d point : points) {
            drawer.drawDot(point, 4, colorPoints, g);
        }
        double shortestDist = Integer.MAX_VALUE;
        for (int i = 0; i < points.length; i++) {
            for (int j = i; j < points.length; j++) {
                if (j != i) {
                    double dist = points[i].getDistanceTo(points[j]);
                    if (dist < shortestDist) shortestDist = dist;
                }
            }
        }
        for (int i = 0; i < points.length; i++) {
            for (int j = i; j < points.length; j++) {
                if (j != i) {
                    if (points[i].getDistanceTo(points[j]) < 1.3 * shortestDist) {
                        g.setColor(colorLines);
                        drawer.drawLine(points[i], points[j], g);
                    }
                }
            }
        }
    }

    @Override
    public void reset() {
        points = new Point3d[points.length];
        forces = new Vec[forces.length];
        for (int i = 0; i < points.length; i++) points[i] = new Point3d(2 * RADIUS * Math.random() - RADIUS, 2 * RADIUS * Math.random() - RADIUS, 2 * RADIUS * Math.random() - RADIUS);
        for (int i = 0; i < forces.length; i++) forces[i] = new Vec(0, 0, 0);
    }
}

File: src/main/java/simulations/PlatonSpace.kt
package simulations

import algebra.Vec3
import framework.Vertex
import framework.WatchDouble
import framework.WatchInt
import framework.physics.PhysicsSimulation
import framework.physics.Sphere
import times
import java.awt.Color
import java.awt.Graphics
import kotlin.concurrent.withLock
import kotlin.math.PI
import kotlin.math.pow

/**
 * Simulation that simulated points that are repelling each other yet can not leave the surface of a sphere.
 */
@Suppress("unused")
class PlatonSpace(amountOfPoint3ds: Int) : PhysicsSimulation("Platon") {
    @WatchInt("Amount of points", 2, 100)
    private var amountOfPoint3ds: Int = amountOfPoint3ds
        set(value) {
            field = value
            while (points.size > value) unregister(tickLock.withLock { points.removeLast() })
            while (points.size < value) {
                val pos = Vec3.random.normalize() * radius
                val sphere = Sphere(pos.x, pos.y, pos.z, sphereSize, 1.0)
                tickLock.withLock { register(sphere) }
                points.add(sphere)
            }
        }

    private val points: ArrayList<Sphere> = ArrayList()
    private val radius = 4.0
    private var colorLines = Vec3(200, 200, 200)
    private var colorPoints = Vec3(163, 153, 239)

    @WatchDouble("Sphere size", 0.01, 3.0)
    private var sphereSize = 1.0
        set(value) {
            field = value
            points.forEach { it.radius = value }
        }

    init {
        g = 0.0;
        frictionPerSecond = 0.0
        reset()
        camera.zoom = 0.001
        camera.x = 0.42
        camera.y = 0.0
        camera.z = 25.0
        camera.phi = 0.0
        camera.theta = PI
    }

    private fun drawAcceleration(g: Graphics) {
        for (i in points.indices) {
            val force = points[i].acceleration
            g.color = Color.ORANGE
            val shorten = 1 / force.length
            camera.renderLine(
                Vertex(points[i].positionVector, colorLines, Vec3.zero),
                Vertex(points[i].positionVector + force * shorten, colorLines, Vec3.zero)
            )
        }
    }

    override fun correctState() {
        points.forEach { it.position = it.position.normalize() * radius }
    }

    override fun calcForces() {
        for (i in points.indices) {
            val first = points[i]
            for (j in i + 1 until points.size) {
                val other = points[j]
                val dif = first.position - other.position
                val dirFirst = (dif - dif.projectOnto(first.position)).normalize()
                val dirSecond = (-dif + dif.projectOnto(other.position)).normalize()
                val scale = 1 / first.getDistanceTo(other).pow(2.0) * 100
                first.applyForce(dirFirst * scale)
                other.applyForce(dirSecond * scale)
            }
        }
    }

    override fun render() {
        super.render()
        val shortestDist = getShortestDist()
        for (i in 0 until points.size) {
            for (j in i until points.size) {
                if (j != i) {
                    try {
                        if (points[i].getDistanceTo(points[j]) < 1.3 * shortestDist) {
                            camera.renderLine(
                                Vertex(points[i].positionVector, colorLines, Vec3.zero),
                                Vertex(points[j].positionVector, colorLines, Vec3.zero)
                            )
                        }
                    } catch (_: IndexOutOfBoundsException) {
                    }
                }
            }
        }
    }

    private fun getShortestDist(): Double {
        var shortestDist = Int.MAX_VALUE.toDouble()
        for (i in 0 until points.size) {
            for (j in i until points.size) {
                if (j != i) {
                    try {
                        val dist = points[i].getDistanceTo(points[j])
                        if (dist < shortestDist) shortestDist = dist
                    } catch (_: IndexOutOfBoundsException) {
                    }
                }
            }
        }
        return shortestDist
    }

    override fun reset() {
        super.reset()
        synchronized(points) {
            points.clear()
            repeat(amountOfPoint3ds) {
                val pos = Vec3.random.normalize() * radius
                val sphere = Sphere(pos.x, pos.y, pos.z, sphereSize, 1.0)
                register(sphere)
                points.add(sphere)
            }
        }
    }
}

fun main() {
    PlatonSpace(4).start()
}